js/app.js

```diff
@@ -0,0 +1,154 @@
+      } else {
+        right.push(array[i]);   // if higher than comparison value then push right
+      }
+    }
+    // update the halves and send them in as the new array when the function is called again.
+    left = quickSort(left);
+    right = quickSort(right);
+    return [].concat(left, popped, right);
+  }
+  
+  console.log(`Quick sort:`);
+  console.log(quickSort(list));
+})();
+
+/*  Merge sort...
+ *  breaks down an array into smaller arrays with a length of 1.
+ *  a number from the left and right side are compared with each pass.
+ *  the lesser number is pushed into result and the greater number is saved for the next comparison.
+ *  this is done until the array is fully reconstructed in numerical order.
+ */
+
+(function myMergeAlgorithm() {
+  const list = makeList(10);
+
+  function mergeSort(array) {
+    if (array.length === 1) {
+      return array;
+    }
+
+    const middle = Math.floor(array.length / 2);
+    const left = array.slice(0, middle);
+    const right = array.slice(middle); // from the middle on
+
+    return merge(
+      mergeSort(left),
+      mergeSort(right),
+    );
+  }
+
+  function merge(left, right) {
+    let result = [ ];
+    let indexLeft = 0;
+    let indexRight = 0;
+
+    while (indexLeft < left.length && indexRight < right.length) {
+      if (left[indexLeft] < right[indexRight]) {
+        result.push(left[indexLeft]);
+        indexLeft++;
+      } else {
+        result.push(right[indexRight]);
+        indexRight++;
+      }
+    }
+
+    return result.concat(left.slice(indexLeft), right.slice(indexRight));
+  }
+
+  console.log(`Merge sort:`);
+  console.log(mergeSort(list));
+})();
+
+// Selection sort
+(function mySelectionAlgorithm() {
+  const list = makeList(10);
+
+  function swap(array, swap, swapWith) {
+    let temp;
+    temp = array[swap];
+    array[swap] = array[swapWith];
+    array[swapWith] = temp;
+  }
+
+  function selectionSort(array) {
+    for (let i = 0; i < array.length - 1; i++) {
+      let minIndex = i; // like the pivot in quick sort, this is your comparison value
+      
+      for (let j = i + 1; j < array.length; j++) {
+        if (array[j] < array[minIndex]) {
+          minIndex = j; // j will always be 1 more than minIndex.
+        }
+      }
+
+      if (minIndex !== i) {
+        // this code is identical. why doesn't it work?
+        // swap(array, array[i], array[minIndex]);
+        temp = array[i];
+        array[i] = array[minIndex];
+        array[minIndex] = temp;
+      }
+    }
+
+    return array;
+  }
+
+  console.log(`Selection sort:`);
+  console.log(selectionSort(list)); // [1, 2, 2, 3, 3, 3, 5, 6, 7]
+})();
+
+// Insertion sort
+(function insertionAlgorithm(array) {
+  const list = makeList(10);
+
+  function insertionSort(array) {
+    for (let i = 1; i < array.length; i++) {
+      let j; // need to define j before the inner loop, otherwise line 146 breaks because the variable j is not being hoisted up.
```

Comment:
Instead of declaring j here, you should just declare a variable that just holds the index you want to swap. Then just set that variable in the inner loop rather than using the j which should be specifically for the for loop.